src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java

```diff
@@ -117,31 +143,79 @@ public void dispose() {
         }
     }
 
-    protected void storeMessages(List<M> messagesList) throws CradleStorageException {
+    @Override
+    public final void handle(T messageBatch) {
+        try {
+            verifyBatch(messageBatch);
+            List<M> messages = getMessages(messageBatch);
+            if (messages.isEmpty()) {
+                if (logger.isWarnEnabled()) {
+                    logger.warn("Empty batch has been received {}", shortDebugString(messageBatch));
+                }
+                return;
+            }
+            M lastMessage = messages.get(messages.size() - 1);
+            long firstSequence = extractSequence(messages.get(0));
+            long lastSequence = extractSequence(lastMessage);
+            SessionKey sessionKey = createSessionKey(lastMessage);
+            SessionBatchHolder holder = sessionToHolder.computeIfAbsent(sessionKey, ignore ->
+                    new SessionBatchHolder(cradleStorage.getObjectsFactory()::createMessageBatch));
+            long prevLastSeq = holder.getAndUpdateSequence(lastSequence);
+            if (prevLastSeq >= firstSequence) {
+                logger.error("Duplicated batch found: {}", shortDebugString(messageBatch));
+                return;
+            }
+            storeMessages(messages, holder);
+        } catch (Exception ex) {
+            logger.error("Cannot handle the batch of type {}", messageBatch.getClass(), ex);
+        }
+    }
+
+    protected void storeMessages(List<M> messagesList, SessionBatchHolder holder) throws CradleStorageException {
         logger.debug("Process {} messages started", messagesList.size());
 
         StoredMessageBatch storedMessageBatch = cradleStorage.getObjectsFactory().createMessageBatch();
         for (M message : messagesList) {
             MessageToStore messageToStore = convert(message);
             storedMessageBatch.addMessage(messageToStore);
         }
-        CompletableFuture<Void> future = store(storedMessageBatch);
-        asyncStoreFutures.put(future, storedMessageBatch);
+        StoredMessageBatch holtBatch;
+        synchronized (holder) {
+            if (holder.add(storedMessageBatch)) {
```

Comment:
Update sequence and add messages aren't executed atomically. Maybe get the last sequence to execute pre-check and encapsulate sequence update into the holder.